### PR TITLE
fixes #26925 - support host registration

### PR DESCRIPTION
**Pull request by timogoebel**
I kept the namespacing for now so we can see if we like it.

### CODE REVIEW

**Comment by ares** [comment]
how comes this didn't get a review yet, perfect candidate for tomorrow triage :-)
[test foreman][test katello][test upgrade]

**Review by ezr-ondrej — COMMENTED** [review]
I am so sorry it totally slipped my mind.
Make sense, works and any improvements should be in followups.
@timogoebel could you please update it from the foreman_register? I'd merge right after.

File: app/services/foreman_register/registration_token.rb
Comment (ekohl) [inline app/services/foreman_register/registration_token.rb:27]: Could we use names? (also true for `prepare_payload`)
```suggestion
      def issued_at
```

File: app/services/foreman_register/registration_token.rb
Comment (ekohl) [inline app/services/foreman_register/registration_token.rb:15]: Should this be at least use constants?

File: app/services/foreman_register/registration_url.rb
Comment (ekohl) [inline app/services/foreman_register/registration_url.rb:15]: Is this a correct interpretation? `require_ssl` states SSL must be used, but IIRC `false` doesn't mean SSL is not available.

File: app/services/foreman_register/registration_url.rb
Comment (ekohl) [inline app/services/foreman_register/registration_url.rb:17]: The API docs are not very clear on whether you need to encode this or if the URI class does this for you. Do you know?

File: app/views/unattended/provisioning_templates/registration/linux_registration_default.erb
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/linux_registration_default.erb:19]: Don't we have a snippet for this? I also think this should be a bit more flexible since this is very much EL7 only.

**Review by ekohl — DISMISSED** [review]

File: app/services/foreman_register/registration_url.rb
Comment (timogoebel) [reply app/services/foreman_register/registration_url.rb:17]: Hm. Does it matter? I believe the token should always be URL safe.

**Comment by timogoebel** [comment]
@ekohl, @ezr-ondrej: Should I keep the namespacing?

File: app/services/foreman_register/registration_url.rb
Comment (ekohl) [reply app/services/foreman_register/registration_url.rb:17]: Hmm, probably ok. I've just gotten used to playing it safe and passing hashes to such libraries which then handle it safely for me so I don't need to think about it. When I pass in strings, I always question it.

**Comment by ekohl** [comment]
> @ekohl, @ezr-ondrej: Should I keep the namespacing?

I don't feel qualified to answer this

**Comment by ezr-ondrej** [comment]
> @ekohl, @ezr-ondrej: Should I keep the namespacing?

I'd say yes for now, I'd like to keep minimal changes. In the ideal world I would say let's remove it in the future, but I feel it's ok if it lives in the namespace for a while.

File: app/views/hosts/_registration_tab.html.erb
Comment (ehelms) [inline app/views/hosts/_registration_tab.html.erb:14]: Could you expand on where you would see this view? I think it can be confusing to think of registering a host that already exists within Foreman and what all that means.

**Comment by ehelms** [comment]
@timogoebel Bear with me, I have two asks :)

 1) Would you mind adding a detailed comment as to the workflow this captures and how the token part works?
 2) I've been working towards an RFC to capture and simplify the variety of client tools and workflows we have across the ecosystem. May I ask that we delay merging this to allow some discussion within the RFC. I promise to include this in that discussion. In a few threads recently we've noted some of the ways we have to ingest hosts. I'd like to expand on that towards simple (and if need be) extenable client features. This PR also hits on an idea being thrown around I planned to include in the RFC around re-using templates rendered from the server for functionality. As it provides parity for provisioning and bootstrapping an existing system.

I'm working to expedite posting said RFC.

File: app/views/hosts/_registration_tab.html.erb
Comment (ekohl) [reply app/views/hosts/_registration_tab.html.erb:14]: I think the idea is that you create a host in Foreman and then it provides the instructions to actually change the host itself by installing/configuring the right tooling etc.

**Comment by timogoebel** [comment]
> @timogoebel could you please update it from the foreman_register? I'd merge right after.

Note to self: This this PR is as up-to-date as it gets.



> 2\. May I ask that we delay merging this to allow some discussion within the RFC.

@ehelms: Sure, not a problem. Let me quickly describe how this PR came to be. At work, we were looking for a tool to manage Linux workstations (e.g. not servers, but computers for actual users).
We compared Foreman and Puppet Enterprise. Puppet Enterprise has a unique feature that allows you to enroll a host by executing a `curl | bash`-style command on the host. This is something we wanted to leverage. Users were asked to set up their computer the way they desired and then execute that particular command. The computer would then check in to the management server, would get a certificate issued and we could then provision wifi configuration and certificates in a secure manner on the device (via puppet). It looked like Puppet Enterprise would have won the comparison, so being a good community member, I decided to add that functionality to Foreman via a plugin.
As I believe the token mechanism can be valuable and there was some interest to have this in core, I decided to open this PR.



> 1. Would you mind adding a detailed comment as to the workflow this captures and how the token part works?

I believe ^^ answers the workflow part. I have to admit, the use case is quite an edge case. It only makes sense with bare-metal hosts that do not use discovery. I'll describe the workflow and my ideas to expand on this in a separate comment.

File: app/controllers/foreman_register/hosts_controller.rb
Comment (timogoebel) [inline app/controllers/foreman_register/hosts_controller.rb:15]: We could extract this to a concern, `Foreman::Controller::AnonymousAccess` or something similar.

File: app/controllers/foreman_register/hosts_controller.rb
Comment (timogoebel) [inline app/controllers/foreman_register/hosts_controller.rb:7]:
```suggestion
      :require_login,
      :check_user_enabled,
```

After #7402 was merged, we need this here as well.

**Comment by ezr-ondrej** [comment]
I do understand the code is not perfect, thats why I'd keep namespacing and improve in follow-ups and remove the namespacing later.

I'd like to give it one more spin and get this merged asap.
Any general objections on having this in core even not perfect @ekohl, @ehelms?

**Comment by timogoebel** [comment]
> I do understand the code is not perfect

Where do you see points that can be improved? Are these code/style issues or workflow related items that can be improved?

**Comment by timogoebel** [comment]
Is this something we potentially want to hide behind a feature toggle?

**Comment by ezr-ondrej** [comment]
I see only the code changes that could be done, as @ekohl and even you (:P) suggested, but I'd try hard not to do more than copy the plugin to core in this PR.
Workflow itself seems quite simple and I myself can't think of any improvements. There I'd let users let us know. Speaking of which would you like to demo it? :upside_down_face: 

> Is this something we potentially want to hide behind a feature toggle?

Given we do not have feature toggles at all, would not be as easy, but we can have a setting to enable/disable this (but setting hiding stuff is not very nice feature toggle IMHO) :thinking:
I'd like the idea of something like rex features globally, but with enable/disable - real feature toggle for users. I'd leave it out of this PR tho

**Comment by ekohl** [comment]
> Is this something we potentially want to hide behind a feature toggle?

This is exactly what I was thinking about. We have settings so that could be the easiest way.

**Comment by timogoebel** [comment]
> 1. Would you mind adding a detailed comment as to the workflow this captures and how the token part works?

The token is a JWT. JWTs can store a signed payload. The only relevant payload that is stored, is the `host_id`.  The token is signed with a secret that is generated for every host. Revoking all tokens for a host simply means changing the secret.
The token is passed as a URL parameter to a controller action that verifies the token is valid, extracts the host_id and renders a template in the context of that host. As it's safe to assume that the request is initiated from that host, the answer can contain secret credentials to obtain a PuppetCA certificate (or a secret per-host activation key if I think about the katello use case).

This could be combined with a lot of things, e.g. a list of hosts could be gathered from a virtualization provider or CMDB. This technique could then be used to setup remote execution to the host. Or to issue a certificate by sending the CSR to Foreman, ...

File: app/controllers/foreman_register/hosts_controller.rb
Comment (ezr-ondrej) [reply app/controllers/foreman_register/hosts_controller.rb:7]: #7402 got merged, but I plan to do some followup anyways. `:check_user_enabled` will do nada if User.current not set, so we don't need this.

**Review by ezr-ondrej — DISMISSED** [review]
Tested, it works. I'm in, @ekohl any major issues?
After (or even before?) merge, we should think of:
- put the registration template in `community_templates`
- documentation

**Comment by ezr-ondrej** [comment]
@timogoebel I've created docu RP: https://github.com/theforeman/foreman-documentation/pull/175, community-templates PR: https://github.com/theforeman/community-templates/pull/734

I believe we are good to merge. The test failures are related though.

**Comment by ehelms** [comment]
In the current implementation, is this only valid for pre-created hosts and via the UI?

**Comment by timogoebel** [comment]
I‘ll fix the test failure next week, currently enjoying a week off.

**Review by tbrisker — COMMENTED** [review]
just a thought - right now, the template installs puppet on the host, which might not be a good fit for users who don't use puppet or who manage puppet outside foreman. I wonder if perhaps we should do just a facter run and post the results to foreman by default, and allow customization of additional setup to run on hosts once they are registered?

**Comment by ekohl** [comment]
Note the `<% if @host.puppetmaster.present? -%>` part in the template. If they don't want Puppet, they shouldn't select a master.

**Comment by timogoebel** [comment]
> In the current implementation, is this only valid for pre-created hosts and via the UI?

@ehelms: Yes.

**Comment by timogoebel** [comment]
Fixed the remaining test.

**Comment by ekohl** [comment]
> > In the current implementation, is this only valid for pre-created hosts and via the UI?
> 
> @ehelms: Yes.

But you can still create hosts via the API and the URL is constant, so it should be easy to script a migration as well, right?

**Comment by timogoebel** [comment]
> But you can still create hosts via the API and the URL is constant, so it should be easy to script a migration as well, right?

Not quite. You can create the hosts via API and I believe this is a great use-case for a brownfield deployment (write a small script that gets a list of hosts from your CMDB, create them in Foreman via hammer or the API). However, the URL of the registration script is only shown in the UI and contains the secret token. It would be easy to expose it via API, but that's something that still needs to be done.

**Comment by tbrisker** [comment]
> You can create the hosts via API and I believe this is a great use-case for a brownfield deployment (write a small script that gets a list of hosts from your CMDB, create them in Foreman via hammer or the API). 

This was exactly the use-case that triggered my previous question - someone on IRC wanted to import their existing hosts into a new foreman, but they don't use any config management solution and don't want all of katello just for sub-man facts. 
If there is some way of leveraging this registration flow for those use cases i think it would be a great win. I'm thinking something along the lines of `curl $foreman/somewhere?maybe_some_token=value | bash` from the host and it will register it to foreman without having to first create the host object in foreman.
Feel free though to tell me if i'm completely misunderstanding the goal of this whole PR and it's unrelated :)

**Comment by ekohl** [comment]
>  If there is some way of leveraging this registration flow for those use cases i think it would be a great win. I'm thinking something along the lines of `curl $foreman/somewhere?maybe_some_token=value | bash` from the host and it will register it to foreman without having to first create the host object in foreman.

Isn't that essentially install `echo "{\"facts\":'$(facter --show-legacy --json)}" | curl -x POST -u myuser https://foreman.example.com/api/hosts/facts` (ugly hack to put it inside facts hash, probably doesn't work)

**Comment by tbrisker** [comment]
> > If there is some way of leveraging this registration flow for those use cases i think it would be a great win. I'm thinking something along the lines of `curl $foreman/somewhere?maybe_some_token=value | bash` from the host and it will register it to foreman without having to first create the host object in foreman.
> 
> Isn't that essentially install `echo "{\"facts\":'$(facter --show-legacy --json)}" | curl -x POST -u myuser https://foreman.example.com/api/hosts/facts` (ugly hack to put it inside facts hash, probably doesn't work)

Pretty much, plus installing facter if needed, wrapping it in a simple to run command and fixing the authentication so the host can actually post to it (i think by default we only accept connections to that endpoints from trusted proxies)

**Comment by timogoebel** [comment]
> Feel free though to tell me if i'm completely misunderstanding the goal of this whole PR and it's unrelated :)

Well, this PR solves one part of the workflow. I'd like to extend it so you can also register hosts that are completely unknown to Foreman. But right now, the host has to exist. Note, that we do have an import mechanism for VMs (and there is a nmap plugin I believe).
What this PR solves is, that you have a mechanism to securely initiate an initial connection between Foreman and a Host without some kind of config management in place.

**Comment by timogoebel** [comment]
> (i think by default we only accept connections to that endpoints from trusted proxies)

Yes, we could issue tokens for that as well. They could already contain the desired loc/org for example.

**Comment by timogoebel** [comment]
[test foreman]

**Comment by timogoebel** [comment]
Ok, tests are 🟢.

**Review by ezr-ondrej — DISMISSED** [review]
Thanks for fixing the test!
I see a lot of ideas and I agree it would be awesome to be able to leverage it.
That sounds to me: let's merge it and continue with follow-ups to make it more useful :)

**Comment by ares** [comment]
I agree with getting this in as is, I think this already unlocks one new interesting flow - import hosts from compute resource and then register them (make them really managed). If this was extended for generic registration in future, that would be really great!

**Comment by timogoebel** [comment]
@ezr-ondrej: Thanks for the fix! :)

**Review by ezr-ondrej — APPROVED** [review]
Thanks @timogoebel!

**Pull request merged**
